tripy/tripy/frontend/trace/ops/pooling.py

```diff
@@ -0,0 +1,157 @@
+)
+def maxpool(
+    input: "tripy.Tensor",
+    kernel_dims: Sequence[int],
+    stride: Sequence[int] = None,
+    padding: Sequence[Tuple[int]] = None,
+) -> "tripy.Tensor":
+    r"""
+    Applies a max pooling over the input tensor.
+
+    The output's non-spatial dimensions are the same as input. For each input spatial dimension
+    :math:`D_{i}`, the corresponding output dimension will be:
+
+    .. math::
+        D_{out_i} = \left\lfloor\frac{D_{i} + \text{padding_before[i]} + \text{padding_after[i]} -
+                \text{kernel_dims[i]}}{\text{stride[i]}} + 1\right\rfloor
+
+    Args:
+        input: The input tensor.
+        kernel_dims: The spatial shape of the pooling window. Only 2-D or 3-D kernel_dims are supported.
+        stride: A sequence of length :math:`M` indicating the stride of pooling across each spatial dimension,
+            where :math:`M` is the number of spatial dimensions, i.e. :math:`M = \text{rank(input)} - 2`.
+            Defaults to all 1.
+        padding: A sequence of pairs of integers of length :math:`M` indicating the zero padding
+            to apply to the input along each spatial dimension before and after the dimension respectively,
+            where :math:`M` is the number of spatial dimensions, i.e. :math:`M = \text{rank(input)} - 2`.
+            Defaults to all 0.
+
+    Returns:
+        The result tensor after the pooling operation.
+
+    .. code-block:: python
+        :linenos:
+        :caption: Example
+
+        input = tp.reshape(tp.arange(16, dtype=tp.float32), (1, 1, 4, 4))
+        output = tp.maxpool(input, kernel_dims=(2, 2))
+
+        pool_torch = torch.nn.MaxPool2d((2, 2), stride=1) # doc: omit
+        expected = pool_torch(torch.from_dlpack(input).to("cpu")) # doc: omit
+
+        assert torch.allclose(torch.from_dlpack(output).to("cpu"), expected)
+    """
+    spatial_dims = len(kernel_dims)
+    if spatial_dims != 2 and spatial_dims != 3:
+        raise_error("Unsupported kernel_dims, must be 2D or 3D.", [f"Got kernel_dims={kernel_dims}"])
+    if stride is not None:
+        if len(stride) != spatial_dims:
+            raise_error(
+                "Stride must have the same length as kernel_dims.",
+                [f"Got stride={stride}, ", f"kernel_dims={kernel_dims}"],
+            )
+    else:
+        stride = [1] * spatial_dims
+    if padding is not None:
+        if len(padding) != spatial_dims:
+            raise_error(
+                "Padding must have the same length as kernel_dims.",
+                [f"Got padding={padding}, ", f"kernel_dims={kernel_dims}"],
+            )
+        if any(len(pad_size) != 2 for pad_size in padding):
+            raise_error(
+                f"Padding must be provided as a sequence of pairs of integers.",
+                details=[f"Supplied padding attribute: {padding} contains sequences that are not of length 2."],
+            )
+    else:
+        padding = [(0, 0)] * spatial_dims
```

Comment:
A lot of this is very similar to what we have for convolution. I'm wondering if we can reuse any of those checks? Unless we think the checks would be different enough that it doesn't help.

Reply:
Refactored this part into a `check_conv_pooling_args` util func.